concurrent_test.go

```diff
@@ -321,3 +384,100 @@ func testResultsDirectory(t *testing.T) string {
 
 	return path
 }
+
+/*
+*********************************************************
+Data structure and functions for analyzing history records
+*********************************************************
+*/
+type workerType string
+
+const (
+	reader workerType = "reader"
+	writer workerType = "writer"
+)
+
+type historyRecord struct {
+	WorkerType workerType `json:"workerType,omitempty"`
+	Key        string     `json:"key,omitempty"`
+	Value      []byte     `json:"value,omitempty"`
+	Txid       int        `json:"txid,omitempty"`
+}
+
+type historyRecords []historyRecord
+
+func (rs historyRecords) Len() int {
+	return len(rs)
+}
+
+func (rs historyRecords) Less(i, j int) bool {
+	// Sorted by key firstly: all records with the same key are grouped together.
+	keyCmp := strings.Compare(rs[i].Key, rs[j].Key)
+	if keyCmp < 0 {
+		return true
+	} else if keyCmp > 0 {
+		return false
+	}
+
+	// Sorted by txid
+	if rs[i].Txid < rs[j].Txid {
+		return true
+	} else if rs[i].Txid > rs[j].Txid {
+		return false
+	}
+
+	// Sorted by workerType: put writer before reader if they have the same txid.
+	if rs[i].WorkerType == writer {
```

Comment:
This is not worker type, but an operation type.

Reply:
updated